src/outputs/histogram.cpp

```diff
@@ -0,0 +1,554 @@
+    PARTHENON_REQUIRE_THROWS(binned_var_component >= 0,
+                             "Negative component indices are not supported");
+  }
+
+  const auto nxbins = x_edges.extent_int(0) - 1;
+  const auto nybins = ndim == 2 ? y_edges.extent_int(0) - 1 : 1;
+
+  result = ParArray2D<Real>(prefix + "result", nybins, nxbins);
+  scatter_result =
+      Kokkos::Experimental::ScatterView<Real **, LayoutWrapper>(result.KokkosView());
+
+  weight_by_vol = pin->GetOrAddBoolean(block_name, prefix + "weight_by_volume", false);
+
+  weight_var_name =
+      pin->GetOrAddString(block_name, prefix + "weight_variable", "HIST_ONES");
+  weight_var_component = -1; // implies that weighting is not applied
+  if (weight_var_name != "HIST_ONES") {
+    weight_var_component =
+        pin->GetInteger(block_name, prefix + "weight_variable_component");
+    // would add additional logic to pick it from a pack...
+    PARTHENON_REQUIRE_THROWS(weight_var_component >= 0,
+                             "Negative component indices are not supported");
+  }
+}
+
+// Computes a 1D or 2D histogram with inclusive lower edges and inclusive rightmost edges.
+// Function could in principle be templated on dimension, but it's currently not expected
+// to be a performance concern (because it won't be called that often).
+void CalcHist(Mesh *pm, const Histogram &hist) {
+  Kokkos::Profiling::pushRegion("Calculate single histogram");
+  const auto x_var_component = hist.x_var_component;
+  const auto y_var_component = hist.y_var_component;
+  const auto binned_var_component = hist.binned_var_component;
+  const auto weight_var_component = hist.weight_var_component;
+  const auto x_var_type = hist.x_var_type;
+  const auto y_var_type = hist.y_var_type;
+  const auto x_edges = hist.x_edges;
+  const auto y_edges = hist.y_edges;
+  const auto x_edges_type = hist.x_edges_type;
+  const auto y_edges_type = hist.y_edges_type;
+  const auto x_edge_min = hist.x_edge_min;
+  const auto x_edge_dbin = hist.x_edge_dbin;
+  const auto y_edge_min = hist.y_edge_min;
+  const auto y_edge_dbin = hist.y_edge_dbin;
+  const auto hist_ndim = hist.ndim;
+  const auto weight_by_vol = hist.weight_by_vol;
+  auto result = hist.result;
+  auto scatter = hist.scatter_result;
+
+  // Reset ScatterView from previous output
+  scatter.reset();
+  // Also reset the histogram from previous call.
+  // Currently still required for consistent results between host and device backends, see
+  // https://github.com/kokkos/kokkos/issues/6363
+  Kokkos::deep_copy(result, 0);
+
+  const int num_partitions = pm->DefaultNumPartitions();
+
+  for (int p = 0; p < num_partitions; p++) {
+    auto &md = pm->mesh_data.GetOrAdd("base", p);
+
+    const auto x_var_pack_string = x_var_type == VarType::Var
+                                       ? std::vector<std::string>{hist.x_var_name}
+                                       : std::vector<std::string>{};
+    const auto x_var = md->PackVariables(x_var_pack_string);
```

Comment:
Not hugely important... but I wonder if it would be best to use the `SparsePack` machinery here? It *might* lead to a performance improvement if you built the `PackDescriptor`s as `static` variables, as that way they are only built once?

Reply:
Probably yes. I just went with what I'm used to and these parts are likely not performance critical anyway.
Do you feel strongly about moving to `SparsePacks`?

Reply:
No I don't feel strongly in this instance---it's not important. I'm mainly just thinking about this from the perspective of eventually moving away from two separate pack infrastructures.